pkg/storage/etcd3/watcher.go

```diff
@@ -182,6 +194,10 @@ func (wc *watchChan) startWatching() {
 			wc.sendEvent(parseEvent(e))
 		}
 	}
+	// When we come to this point, it's possible that:
+	// - client side ends the watch, e.g. cancel the context, close the client.
+	// - server side ends the watch.
+	close(watchClosedCh)
```

Comment:
do this via defer if it should always happen. right now there's a case where it isn't called and I can't tell if that's bug or not.